### PR TITLE
Webhook for blocking pods in namespaces openshift-logging, openshift-operators and non-core mnamespaces from deploying on master and infra.

### PR DESCRIPTION
**Pull request by aliceh**
This commit adds in pod.go a webhook that will check the namespace in which the pods deploy, and if that namespace is openshift-logging, openshift-operators or a non-core namespace the deployment will not be allowed on master and infra nodes. This commit prevents the logging stack from targeting the infra nodes [OSD-4262](https://issues.redhat.com/browse/OSD-5269) and prevents customers from launching pods with workloads on master and infra JIRA ticket [OSD-4333](https://issues.redhat.com/browse/OSD-4333).
The unit tests for the pod webhook are in pod_test.go.

### CODE REVIEW

**Comment by lisa** [comment]
/hold


for review

File: pkg/webhooks/pod/pod.go
Comment (cblecker) [inline pkg/webhooks/pod/pod.go:63]: This should have a comment about why we are ignoring it, and how important it is that this doesn't change.

**Review by lisa — CHANGES_REQUESTED** [review]
Hey @aliceh great start! You're missing a way to register the webhook, look at the other `add_*.go` files for inspiration. Once that file is in place you need to `make syncset` and commit the new `build/selectorsyncset.yaml` file.

I'd also like to see a test case for infra node and noschedule taint, please.

/cc @cblecker as well.

File: pkg/webhooks/pod/pod_test.go
Comment (aliceh) [reply pkg/webhooks/pod/pod_test.go:111]: my understanding was that the non-core namespaces should be blocked from deploying pods on master/infra when there's toleration for NoSchedule taint.

File: pkg/webhooks/pod/pod.go
Comment (aliceh) [reply pkg/webhooks/pod/pod.go:22]: The privileged namespaces are the ones that are able to deploy, we are excluding from that set the exception namespaces openshift-logging and openshift-operators which we block from deploying, same as non-privileged customer namespaces.

File: pkg/webhooks/pod/pod.go
Comment (lisa) [inline pkg/webhooks/pod/pod.go:112]:
```suggestion
	// If the incoming Pod is aimed at a privileged namespace except for exceptionNamespace, allow it to do whatever it wants.
	// However, if the pod is targeting a customer's namespace (aka non-privileged), then it may not tolerate certain master/infra node taints.
```

File: pkg/webhooks/pod/pod.go
Comment (lisa) [inline pkg/webhooks/pod/pod.go:113]: This is a fairly complex conditional and it'd be clearer to express what is being matched if it were refactored into a well-named function. Perhaps `isRequestPrivileged` or `isRequestUnprivileged`.

File: pkg/webhooks/pod/pod.go
Comment (lisa) [inline pkg/webhooks/pod/pod.go:139]:
```suggestion
	ret = admissionctl.Allowed("Allowed to create Pod because of RBAC")
```

Here it will apply for all namespaces, not just privileged ones.

File: pkg/webhooks/pod/pod_test.go
Comment (lisa) [inline pkg/webhooks/pod/pod_test.go:16]:
```suggestion
```

Extra newline

File: pkg/webhooks/pod/pod.go
Comment (lisa) [inline pkg/webhooks/pod/pod.go:144]: There will need to be some refactor once #105 is merged.

**Review by lisa — CHANGES_REQUESTED** [review]
Some feedback.

I'll continue with unit test review, but my day is nearly over and I wanted to get this to you in time.

File: pkg/webhooks/pod/pod.go
Comment (aliceh) [reply pkg/webhooks/pod/pod.go:22]: My reasoning behind the choice of name was that we have the core namespaces that are privileged as a group in other contexts too for example in the case of RBAC subject permissions, so the complement of that group would be the unprivileged namespaces. In this context, it seemed reasonable to call the set of two namespaces that we want to exclude from the privileged set exceptions. Of course, I am happy to rename that set unprivileged if you feel that it's better.

**Comment by lisa** [comment]
@cblecker Looks okay to me. What are your thoughts?

@aliceh Can you squash the commits into one?

**Comment by lisa** [comment]
/lgtm
/hold

Please let @csheremeta or @cblecker remove my hold since I am out for four days and #109 needs to take priority for promoting.

**Comment by lisa** [comment]
@aliceh @cblecker Is this ready to merge still?

**Comment by jewzaam** [comment]
Only going by the title, but is this not limiting Projects from deploying to master / infra?

**Comment by jewzaam** [comment]
> Only going by the title, but is this not limiting Projects from deploying to master / infra?

I didn't notice the "non-core"  mention in the title , my bad.

**Review by lisa — APPROVED** [review]

**Pull request merged**
